docs/console-access.md

```diff
@@ -0,0 +1,340 @@
+
+Choice of access control methods. Either QEMU can mandate client certificates
+and perform access control checks on the certificate identity (planned, but not
+yet supported in QEMU & libvirt), or protocol specific authentication must be
+used.
+
+Only supports VNC and SPICE, until QEMU is able to do fine grained ACL checks
+against client certs.
+
+
+### 2. HTTP CONNECT proxy
+
+QEMU on the compute node running a TCP socket with TLS and x509 certificates
+enabled.
+
+Compute node certs must be issued by a CA that is trusted by all tenant
+users and new certs acquired for each new compute node.
+
+One or more proxy servers able to multiplex to an arbitrary number of backend
+compute nodes
+
+Tenant users make HTTP CONNECT request to proxy server, giving the IP address
+and port on the cluster network that QEMU is listening on
+
+Tenant users directly validate the received server certificates against
+the requested IP + port of the QEMU service.
+
+Choice of access control methods. Either QEMU can mandate client certificates
+and perform access control checks on the certificate identity (planned, but not
+yet supported in QEMU & libvirt), or protocol specific authentication must be
+used.
+
+Only supports VNC and SPICE, until QEMU is able to do fine grained ACL checks
+against client certs.
+
+
+### 3. HTTP websockets proxy
+
+QEMU on the compute node running a TCP socket with TLS and x509 certificates
+enabled.
+
+Compute node certs can be issued by a private CA, which only needs to be
+trusted by the proxy servers. Proxies are issued client certs by the CA
+for mutual auth.
+
+One or more proxy servers able to multiplex to an arbitrary number of backend
+compute nodes
+
+Shared secret between the proxy and tenant user identifies the QEMU service to
+connect to and authenticates the user.
+
+Tenant users validate the received server certificates against the IP + port
+of the proxy server.
+
+The proxy server performs TLS handshake with the QEMU servers and validates its
+certificate.
+
+Protocol specific authentication is not exposed to tenant users.
+
+Supports VNC, SPICE and serial consoles with parity of functionality.
+
+
+Proxy Implementation
+--------------------
+
+For both the HTTP CONNECT and HTTP websockets proxy there needs to be some be some
+intelligence in the proxy. In the HTTP CONNECT proxy, there needs to be code to
+validate that the requested IP+port pair actually corresponds to a valid guest
+console connection. This protects against the proxy being used to attack other
+arbitrary network servers running in the cluster network. In the HTTP websockets
+proxy, there needs to be code to validate the shared secret obtained from the
+client, to resolve it into a IP+port for the QEMU connection. It also needs to
+be prepared to do some intelligent MITM of the connections to handle authentication
+and migration.
+
+The libvirt-console-proxy project aims to provide a general purpose server that
```

Comment:
I think in the context, how kubernetes works (apiserver has the authority over the data and it's views), I would rather see the resolver endpoint in the apiserver. I am pretty sure, that for such things, you would not see an extra component in Kubernetes for that. If we really want extra containers for such small things, I am not sure that this will lead to a nice and maintainable architecture.

A good example are Kubernetes Ingress controllers. There are for instance implementations using haproxy and nginx. In order to provide allowed routes to them, the apiserver has an extra endpoints API, which allows you to get all the routing information, for all services belonging to a specific app or task. If the proxy pod is taken over, the user can query the only that information.

I have also thought about more ingress-style like integration for such websocket endoints. It would be possible to create headless services for VMs, which expose spice and vnc endpoints. One plus (and minus) would be that you could provide the whole routing information via a controller, and don't have a resolve lookup. If we would have some general access control on the proxy, that might also be interesting. I am not saying that we should pursue that ingress-style approach, just to share the thoughts ...